File: app/src/main/java/me/instabattle/app/activities/BattleActivity.java
package me.instabattle.app.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ListView;
import android.widget.TextView;

import java.util.List;

import me.instabattle.app.R;
import me.instabattle.app.services.LocationService;
import me.instabattle.app.settings.State;
import me.instabattle.app.adapters.EntryListAdapter;
import me.instabattle.app.models.Entry;
import me.instabattle.app.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BattleActivity extends Activity {

    private static final String TAG = "BattleActivity";

    private EntryListAdapter entryListAdapter;
    private ListView entryList;

    public static Class<?> gotHereFrom;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_battle);

        ((TextView) findViewById(R.id.battle_title)).setText(State.chosenBattle.getName());
        ((TextView) findViewById(R.id.battle_description)).setText(State.chosenBattle.getDescription());

        entryList = (ListView) findViewById(R.id.entryList);

        State.chosenBattle.getEntriesAndDo(new Callback<List<Entry>>() {
            @Override
            public void onResponse(Call<List<Entry>> call, Response<List<Entry>> response) {
                entryListAdapter = new EntryListAdapter(BattleActivity.this, response.body());
                entryList.setAdapter(entryListAdapter);
            }

            @Override
            public void onFailure(Call<List<Entry>> call, Throwable t) {
                Utils.showToast(BattleActivity.this, "Failed to get battle entries, try again later.");
                Log.e(TAG, "cant get entries: " + t);
            }
        });
    }

    @Override
    public void onBackPressed() {
        Intent back = new Intent(this, gotHereFrom);
        startActivity(back);
    }

    public void vote(View v) {
        if (State.chosenBattle.getEntriesCount() > 1) {
            State.chosenBattle.getVoteAndDo(new Callback<List<Entry>>() {
                @Override
                public void onResponse(Call<List<Entry>> call, Response<List<Entry>> response) {
                    if (response.code() == 400) {
                        Utils.showToast(BattleActivity.this, "You've already voted for all pairs of photos.");
                        return;
                    }

                    VoteActivity.firstEntry = response.body().get(0);
                    VoteActivity.secondEntry = response.body().get(1);

                    Intent voting = new Intent(BattleActivity.this, VoteActivity.class);
                    startActivity(voting);

                    Log.d(TAG, "got vote");
                }

                @Override
                public void onFailure(Call<List<Entry>> call, Throwable t) {
                    Utils.showToast(BattleActivity.this, "Failed to set new vote, try again later.");
                    Log.e(TAG, "cant get vote: " + t);
                }
            });
        } else {
            Utils.showToast(BattleActivity.this, "Only one entry in battle, you can't vote.");
        }
    }

    public void participate(View v) {
        if (!LocationService.hasActualLocation()) {
            Utils.showToast(BattleActivity.this, "There're problems with detecting your location. Try again later.");
        } else if (LocationService.isTooFarFrom(State.chosenBattle)) {
            Utils.showToast(BattleActivity.this, "You're too far away, come closer to battle for participating!");
        } else {
            CameraActivity.gotHereFrom = BattleActivity.class;
            Intent participating = new Intent(this, CameraActivity.class);
            participating.putExtra("battleTitle", State.chosenBattle.getName());
            startActivity(participating);
        }
    }
}

File: app/src/main/kotlin/me/instabattle/app/activities/BattleActivity.kt
package me.instabattle.app.activities

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import kotlinx.android.synthetic.main.activity_battle.*
import me.instabattle.app.R
import me.instabattle.app.adapters.EntryListAdapter
import me.instabattle.app.models.Entry
import me.instabattle.app.services.LocationService
import me.instabattle.app.settings.GlobalState
import org.jetbrains.anko.error
import org.jetbrains.anko.info
import org.jetbrains.anko.startActivity
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class BattleActivity : DefaultActivity() {
    private lateinit var recyclerView: RecyclerView
    private lateinit var viewAdapter: RecyclerView.Adapter<*>
    private lateinit var viewManager: RecyclerView.LayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_battle)

        viewManager = LinearLayoutManager(this)

        battle_title.text = GlobalState.chosenBattle!!.name
        battle_description.text = GlobalState.chosenBattle!!.description

        GlobalState.chosenBattle!!.getEntriesAndDo(object : Callback<List<Entry>> {
            override fun onResponse(call: Call<List<Entry>>, response: Response<List<Entry>>) {
                viewAdapter = EntryListAdapter(this@BattleActivity, response.body()!!)


                recyclerView = findViewById<RecyclerView>(R.id.entryList).apply {
                    setHasFixedSize(true)
                    layoutManager = viewManager
                    adapter = viewAdapter
                }
            }

            override fun onFailure(call: Call<List<Entry>>, t: Throwable) {
                toast("Failed to get battle entries, try again later.")
                error { "cant get entries: $t" }
            }
        })
    }

    override fun onResume() {
        super.onResume()

        if (::viewAdapter.isInitialized)
            viewAdapter.notifyDataSetChanged()
    }

    override fun onBackPressed() {
        val back = Intent(this, gotHereFrom)
        startActivity(back)
    }

    fun vote(v: View) {
        if (GlobalState.chosenBattle!!.entriesCount > 1) {
            GlobalState.chosenBattle!!.getVoteAndDo(object : Callback<List<Entry>> {
                override fun onResponse(call: Call<List<Entry>>, response: Response<List<Entry>>) {
                    if (response.code() == 400) {
                        toast("You've already voted for all pairs of photos.")
                        return
                    }

                    startActivity<VoteActivity>(
                            "firstEntry" to response.body()!![0],
                            "secondEntry" to response.body()!![1]
                    )

                    info { "got vote" }
                }

                override fun onFailure(call: Call<List<Entry>>, t: Throwable) {
                    toast("Failed to set new vote, try again later.")
                    error { "cant get vote: $t" }
                }
            })
        } else {
            toast("Only one entry in battle, you can't vote.")
        }
    }

    fun participate(v: View) {
        if (!LocationService.hasActualLocation()) {
            toast("There're problems with detecting your location. Try again later.")
        } else if (LocationService.isTooFarFrom(GlobalState.chosenBattle)) {
            toast("You're too far away, come closer to battle for participating!")
        } else {
            startActivity<CameraViewActivity>()
        }
    }

    companion object {
        var gotHereFrom: Class<*>? = null
    }
}
